include/seqan3/search/algorithm/search_result_range.hpp

```diff
@@ -0,0 +1,320 @@
+    //!\brief Pointer type is pointer of container element type.
+    using pointer = std::add_pointer_t<value_type>;
+    //!\brief Sets iterator category as input iterator.
+    using iterator_category = std::input_iterator_tag;
+    //!\}
+
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    constexpr search_result_range_iterator() noexcept = default; //!< Defaulted.
+    constexpr search_result_range_iterator(search_result_range_iterator const &) noexcept = default; //!< Defaulted.
+    constexpr search_result_range_iterator(search_result_range_iterator &&) noexcept = default; //!< Defaulted.
+    constexpr search_result_range_iterator & operator=(search_result_range_iterator const &) noexcept = default; //!< Defaulted.
+    constexpr search_result_range_iterator & operator=(search_result_range_iterator &&) noexcept = default; //!< Defaulted.
+    ~search_result_range_iterator() = default; //!< Defaulted.
+
+    /*!\brief Construct from the associated search result range.
+     * \param[in] range The associated search result range.
+     * \param[in] first_call_of_begin A flag indicating whether this is the first call of begin.
+     *
+     * \details
+     *
+     * Initialises the iterator with the associated search result range. If `first_call_of_begin` is `true`,
+     * the search algorithm is invoked on the first query.
+     */
+    constexpr search_result_range_iterator(search_result_range & range, bool const first_call_of_begin) :
+        range_ptr{& range},
+        at_end(range.at_end())
+    {
+        if (first_call_of_begin)
+            fetch_next_query_results();
+    }
+    //!\}
+
+    /*!\name Access operators
+     * \{
+     */
+
+    /*!\brief Access the pointed-to element.
+     * \returns A reference to the current element.
+     */
+    //!\brief Returns the current search result.
+    reference operator*() const noexcept
+    {
+        return *current_buffer_it;
+    }
+
+    //!\brief Returns a pointer to the current search result.
+    pointer operator->() const noexcept
+    {
+        return std::addressof(*current_buffer_it);
+    }
+    //!\}
+
+    /*!\name Increment operators
+     * \{
+     */
+    //!\brief Increments the iterator by one.
+    search_result_range_iterator & operator++(/*pre*/)
+    {
+        assert(range_ptr != nullptr);
+
+        if (++current_buffer_it == last_buffer_it)
+            fetch_next_query_results();
+
+        return *this;
+    }
+
+    //!\brief Increments the iterator by one.
+    void operator++(int /*post*/)
+    {
+        ++(*this);
+    }
+    //!\}
+
+    /*!\name Comparison operators
+     * \{
+     */
+    //!\brief Checks whether `lhs` is equal to the sentinel.
+    friend constexpr bool operator==(search_result_range_iterator const & lhs,
+                                     std::ranges::default_sentinel_t const &) noexcept
+    {
+        return lhs.at_end && (lhs.current_buffer_it == lhs.last_buffer_it);
```

Comment:
It does work, once I befriend `operator==()` in the enclosing range class. Note, that friendship is not transitive. This means just because `friend operator==()` can access the private member of my_range::iterator it cannot access the private member of the enclosing class my_range. Accordingly, the `operator==` cannot access private members of my_range through a pointer stored in my_range::iterator. Here is an example: https://godbolt.org/z/ZZuRfk

Reply:
But it is not worht the effort to add complicated friend declarations just to safe a bool in a move-only iterator.

Reply:
> But it is not worht the effort to add complicated friend declarations just to safe a bool in a move-only iterator.

Yes I agree. I just didn't know that befriending is not transitive, thanks for the explanation.